spatialprofilingtoolbox/apiserver/app/main.py

```diff
@@ -176,20 +174,38 @@ async def request_phenotype_proximity_computation(
     retrieve = query().retrieve_signature_of_phenotype
     criteria1 = retrieve(phenotype1, study)
     criteria2 = retrieve(phenotype2, study)
-    with CountRequester() as requester:
+    with OnDemandRequester() as requester:
         metrics = requester.get_proximity_metrics(
             query().get_study_components(study).measurement,
             radius,
-            [
+            (
                 criteria1.positive_markers,
                 criteria1.negative_markers,
                 criteria2.positive_markers,
                 criteria2.negative_markers,
-            ],
+            ),
         )
     return metrics
 
 
+@app.get("/request-squidpy-computation/")
+async def request_squidpy_computation(
+    study: ValidStudy,
+    phenotypes: list[ValidPhenotype],
```

Comment:
This does indeed generalize the Proximity calculation call signature, but I don't think this is the model we want to emulate in the case of the phenotype arguments. Rather, the model is the Counts calculation, since that supports user-defined (arbitrary) composite phenotype definitions.
If we were limiting ourselves only to the pre-existing phenotype definitions plus single channels, then we really could pre-calculate everything with no need for the ondemand functionality.

Reply:
(We would like this for proximity metrics as well, but we just haven't developed the UI.)